pkg/controller/operators/olm/operatorgroup.go

```diff
@@ -315,54 +331,92 @@ func (a *Operator) ensureSingletonRBAC(operatorNamespace string, csv *v1alpha1.C
+	targetCSV, err := a.lister.OperatorsV1alpha1().ClusterServiceVersionLister().ClusterServiceVersions(targetNamespace).Get(csv.GetName())
+	if err != nil {
+		return err
+	}
+
 	ownerSelector := ownerutil.CSVOwnerSelector(csv)
 	ownedRoles, err := a.lister.RbacV1().RoleLister().Roles(operatorNamespace).List(ownerSelector)
 	if err != nil {
 		return err
 	}
 
-	for _, r := range ownedRoles {
+	targetRoles, err := a.lister.RbacV1().RoleLister().List(ownerutil.CSVOwnerSelector(targetCSV))
+	if err != nil {
+		return err
+	}
+
+	targetRolesByName := map[string]*rbacv1.Role{}
+	for _, r := range targetRoles {
+		targetRolesByName[r.GetName()] = r
+	}
+
+	for _, ownedRole := range ownedRoles {
 		// don't trust the owner label
-		if !ownerutil.IsOwnedBy(r, csv) {
+		if !ownerutil.IsOwnedBy(ownedRole, csv) {
 			continue
 		}
-		_, err := a.lister.RbacV1().RoleLister().Roles(targetNamespace).Get(r.GetName())
-		if err != nil {
-			r.SetNamespace(targetNamespace)
-			if _, err := a.OpClient.CreateRole(r); err != nil {
+
+		existing, ok := targetRolesByName[ownedRole.GetName()]
+
+		// role already exists, update the rules
+		if ok {
+			existing.Rules = ownedRole.Rules
+			if _, err := a.OpClient.UpdateRole(existing); err != nil {
 				return err
 			}
+			continue
+		}
+
+		// role doesn't exist, create it
+		ownedRole.SetNamespace(targetNamespace)
+		ownedRole.SetOwnerReferences([]metav1.OwnerReference{ownerutil.NonBlockingOwner(targetCSV)})
+		if _, err := a.OpClient.CreateRole(ownedRole); err != nil {
+			return err
 		}
-		// TODO check rules
 	}
 
 	ownedRoleBindings, err := a.lister.RbacV1().RoleBindingLister().RoleBindings(operatorNamespace).List(ownerSelector)
 	if err != nil {
 		return err
 	}
 
+	targetRoleBindings, err := a.lister.RbacV1().RoleBindingLister().List(ownerutil.CSVOwnerSelector(targetCSV))
+	if err != nil {
+		return err
+	}
+
+	targetRoleBindingsByName := map[string]*rbacv1.RoleBinding{}
+	for _, r := range targetRoleBindings {
+		targetRoleBindingsByName[r.GetName()] = r
+	}
+
 	// role bindings
-	for _, r := range ownedRoleBindings {
+	for _, ownedRoleBinding := range ownedRoleBindings {
 		// don't trust the owner label
-		if !ownerutil.IsOwnedBy(r, csv) {
+		if !ownerutil.IsOwnedBy(ownedRoleBinding, csv) {
 			continue
 		}
-		_, err := a.lister.RbacV1().RoleBindingLister().RoleBindings(targetNamespace).Get(r.GetName())
-		if err != nil {
-			r.SetNamespace(targetNamespace)
+		_, ok := targetRolesByName[ownedRoleBinding.GetName()]
 
-			if _, err := a.OpClient.CreateRoleBinding(r); err != nil {
-				return err
-			}
-			// TODO check rules
+		// role binding exists
+		if ok {
+			// TODO: should we check if SA/role has changed?
+			continue
+		}
+
+		// role binding doesn't exist
+		ownedRoleBinding.SetNamespace(targetNamespace)
+		ownedRoleBinding.SetOwnerReferences([]metav1.OwnerReference{ownerutil.NonBlockingOwner(targetCSV)})
```

Comment:
Nice, much better to set the `ownerReferences` to the copied CSV in that namespace.